albatross/models/ransac.h

```diff
@@ -127,143 +121,128 @@ ransac(const typename RansacFunctions<FitType>::Fitter &fitter,
+                                .mean();
+        return mean_score;
       };
 
   const auto best_inds = ransac<FitType>(
       fitter, inlier_metric, model_metric, map_values(fold_indexer),
       inlier_threshold, random_sample_size, min_inliers, max_iterations);
-  RegressionDataset<FeatureType> best_dataset(
-      subset(best_inds, dataset.features), subset(best_inds, dataset.targets));
-  return best_dataset;
+  return subset(dataset, best_inds);
 }
 
+template <typename ModelType, typename MetricType, typename FeatureType>
+struct Fit<Ransac<ModelType, MetricType>, FeatureType> {
+
+  using FitModelType = typename fit_model_type<ModelType, FeatureType>::type;
+
+  Fit(const FitModelType &fit_model_) : fit_model(fit_model_){};
+
+  FitModelType fit_model;
+};
+
 /*
- * This wraps any other RegressionModel and performs ransac each time fit is
- * called.
- *
- * Note that the model pointer passed into the constructor is NOT const and will
- * be updated any time `fit` is called on this class.
+ * This wraps any other model and performs ransac each time fit is called.
  */
-template <typename ModelType, typename FeatureType>
-class GenericRansac : public RegressionModel<FeatureType> {
+template <typename ModelType, typename MetricType>
+class Ransac : public ModelBase<Ransac<ModelType, MetricType>> {
 public:
-  GenericRansac(ModelType *sub_model, double inlier_threshold,
-                std::size_t min_inliers, std::size_t random_sample_size,
-                std::size_t max_iterations,
-                const IndexerFunction<FeatureType> &indexer_function =
-                    leave_one_out_indexer<FeatureType>)
-      : sub_model_(sub_model), inlier_threshold_(inlier_threshold),
-        min_inliers_(min_inliers), random_sample_size_(random_sample_size),
-        max_iterations_(max_iterations), indexer_function_(indexer_function),
-        metric_(evaluation_metrics::negative_log_likelihood){};
-
-  std::string get_name() const override {
-    std::ostringstream oss;
-    oss << "ransac[" << sub_model_->get_name() << "]";
-    return oss.str();
+  Ransac(const ModelType &sub_model, const MetricType &metric,
+         double inlier_threshold, std::size_t min_inliers,
+         std::size_t random_sample_size, std::size_t max_iterations)
+      : sub_model_(sub_model), metric_(metric),
+        inlier_threshold_(inlier_threshold), min_inliers_(min_inliers),
+        random_sample_size_(random_sample_size),
+        max_iterations_(max_iterations){};
+
+  static_assert(
+      std::is_base_of<EvaluationMetric<Eigen::VectorXd>, MetricType>::value ||
+          std::is_base_of<EvaluationMetric<MarginalDistribution>,
+                          MetricType>::value ||
+          std::is_base_of<EvaluationMetric<JointDistribution>,
+                          MetricType>::value,
+      "MetricType must be an EvaluationMetric.");
+
+  std::string get_name() const {
+    return "ransac[" << sub_model_->get_name() << "]";
```

Comment:
Does this actually work? Shouldn't this be using a `std::stringstream`?

Reply:
heh, good catch.  I actually just fixed that in a subsequent commit, I'll bring it into this one.